Math/NumberTheory/Moduli/CubicSymbol.hs

```diff
@@ -0,0 +1,125 @@
+  -- In particular, it returns an error if @beta == 0@
+  0 -> error "Math.NumberTheory.Moduli.CubicSymbol: denominator is not coprime to 3."
+  _ -> cubicSymbolHelper alpha beta
+  where
+    betaNorm = norm beta
+
+cubicSymbolHelper :: EisensteinInteger -> EisensteinInteger -> CubicSymbol
+cubicSymbolHelper alpha beta = cubicReciprocity primaryRemainder primaryBeta <> newSymbol
+  where
+    (primaryRemainder, primaryBeta, symbolExponent) = extractPrimaryContributions remainder beta
+    remainder = A.rem alpha beta
+    newSymbol = exponentiation unmodularExponent Omega
+    unmodularExponent = wordToInt (unMod symbolExponent)
+    exponentiation k x = if k == 0 then One else stimes k x
+
+-- This function first checks if its arguments are zeros or units. If they are not,
+-- it invokes cubic reciprocity by calling cubicSymbolHelper with swapped arguments.
+cubicReciprocity :: EisensteinInteger -> EisensteinInteger -> CubicSymbol
+-- Note @cubicReciprocity 0 1 = One@. It turns out it is better to adopt this convention.
+cubicReciprocity _ 1 = One
+-- Checks if first argument is zero. Note the second argument is never zero.
+cubicReciprocity 0 _ = Zero
+-- This checks if the first argument is a unit. Because it's primary,
+-- it is enough to pattern match with 1.
+cubicReciprocity 1 _ = One
+-- Otherwise, cubic reciprocity is called.
+cubicReciprocity alpha beta = cubicSymbolHelper beta alpha
+
+-- This function takes two Eisenstein intgers @alpha@ and @beta@ and returns three
+-- arguments @(gamma, delta, contribution)@. @gamma@ and @delta@ are the associated
+-- primary numbers to alpha and beta respectively. @contribution@ is a an integer
+-- defined mod 3 which measures the difference between the cubic residue of @alpha@
+-- and @beta@ with respect to the cubic residue of @gamma@ and @delta@.
+extractPrimaryContributions :: EisensteinInteger -> EisensteinInteger -> (EisensteinInteger, EisensteinInteger, Mod 3)
+extractPrimaryContributions alpha beta = (gamma, delta, contribution)
+  where
+    contribution = j*m - i*m -i*n
+    [i, j, m, n] = map fromIntegral [iInt, jInt, mInt, nInt]
+    mInt :+ nInt = A.quot (delta - 1) 3
+    (iInt, gamma) = getPrimaryDecomposition alphaThreeFree
+    (_, delta) = getPrimaryDecomposition beta
+    jInt = wordToInteger jIntWord
+    -- This function outputs data such that
+    -- @(1 - ω)^jIntWord * alphaThreeFree = alpha@.
+    (jIntWord, alphaThreeFree) = splitOff (1 - ω) alpha
+
+-- This function takes an Eisenstein number and returns its primary decomposition @(powerUnit, factor)@
+-- That is, given @e@ coprime with 3, it returns a unique integer (mod 6) @powerUnit@ and a unique
+-- Eisenstein number @factor@ such that @(1 + ω)^powerUnit * e = 1 + 3*factor@.
+-- Note that L.findIndex cannot return Nothing. This happens only if @e@ is not
+-- coprime with 3. This cannot happen since @U.splitOff@ is called just before.
+getPrimaryDecomposition :: EisensteinInteger -> (Integer, EisensteinInteger)
```

Comment:
Could you possibly return `(CubicSymbol, EisensteinInteger)` and make corresponding changes to `extractPrimaryContributions` and `cubicHelper`? This will be more natural representation than `Integer` and `Mod 3`, I think. 

I suggest to pattern match on possible values of `remainder` and return corresponding cubic symbols directly.

Reply:
This is probably better

Reply:
I changed it but I am disappointed by the result. The code looks messier and it runs more slowly. I thought about it but I could not find any better way of implementing these changes. If you have any suggestions to improve it, let me know.

Reply:
That's surprising, I'll take a look next week. I should revisit `M.NT.Q.EisensteinIntegers` to check that all operations are as efficient as possible. 

In the meantime I have three suggestions:

* Instead of relying on autoderived `stimes` (which appears to be awkward) we can actually define our own, supporting zeros and negatives. This basically brings us back to your very first design, sorry for the digression.

```haskell
instance Semigroup CubicSymbol where
  (<>) = ... 
  stimes k n = case (k `mod` 3, n) of ...
```

* Could you please implement `symbolToNum :: CubicSymbol -> EisensteinInteger`?

Reply:
I did this. In the new commit, the timings went down to the previous level.